### ISSUE TITLE
[question] How to use package revisions with short windows paths?

### ISSUE THREAD

**Issue by ls-nano** [conan-io/conan#14400]
### How to use package revisions with short windows paths?

Hello,

we are currently in the process of enabling package revisions for our Conan CI builds. This works great on most platforms, however, on windows we run into some problems, especially regarding windows short paths. We run Conan 1.60.1.

We are currently running conan in short paths mode as we expierenced issues when running it with long paths. After we've built the package, we then go through all `.conan_link` files and copy the contents of the directory the link points to into the actual package. This is so that we can upload and move the package around without having to fear breaking the links. So far this has worked great, but since enabling revisions `conan search` does not find the packages anymore (it does before moving the files).  This prevents us from uploading the package. Is there a way to combine revisions and our short path building approach?  

I _suspect_ this is due to revisions creating a hash and `conan search` checking the saved hash against the found package, and the two hashes diverging after the files were moved, however, I'm puzzled why that would be as I would _expect_ conan to follow the link in the `.conan_link` file and hash the directory it is pointing to, which _should_ be the same hash as our approach should produce? Or am I misunderstanding something here?   

Our profile is: 
```
[settings]
os=Windows
compiler="Visual Studio"
compiler.version=16
compiler.runtime=MDd
arch=x86_64
os_build=Windows
arch_build=x86_64

[env]
# Target Windows 7
# https://docs.conan.io/en/latest/reference/build_helpers/visual_studio.html#visual-studio-build
CL=/D_WIN32_WINNT=0x0601 /DWINVER=0x0601
# CMake should pick up on this
CFLAGS=/D_WIN32_WINNT=0x0601 /DWINVER=0x0601
# When trying to build m4, there were errors about missing C compiler.
# This is probably an issue for anything that uses msys2 (https://www.msys2.org/docs/what-is-msys2/)
# Having this variable set fixes it
CC_FOR_BUILD=cl
```

and our conan.conf is:
```
[log]
run_to_output = True        # environment CONAN_LOG_RUN_TO_OUTPUT
run_to_file = False         # environment CONAN_LOG_RUN_TO_FILE
level = critical            # environment CONAN_LOGGING_LEVEL
# trace_file =              # environment CONAN_TRACE_FILE
print_run_commands = False  # environment CONAN_PRINT_RUN_COMMANDS

[general]
default_profile = default
compression_level = 9                 # environment CONAN_COMPRESSION_LEVEL
sysrequires_sudo = True               # environment CONAN_SYSREQUIRES_SUDO
request_timeout = 60                  # environment CONAN_REQUEST_TIMEOUT (seconds)
default_package_id_mode = semver_direct_mode # environment CONAN_DEFAULT_PACKAGE_ID_MODE
# parallel_download = 8               # experimental download binaries in parallel
# full_transitive_package_id = 0
# retry = 2                           # environment CONAN_RETRY
# retry_wait = 5                      # environment CONAN_RETRY_WAIT (seconds)
# sysrequires_mode = enabled          # environment CONAN_SYSREQUIRES_MODE (allowed modes enabled/verify/disabled)
# vs_installation_preference = Enterprise, Professional, Community, BuildTools # environment CONAN_VS_INSTALLATION_PREFERENCE
# verbose_traceback = False           # environment CONAN_VERBOSE_TRACEBACK
# error_on_override = False           # environment CONAN_ERROR_ON_OVERRIDE
# bash_path = ""                      # environment CONAN_BASH_PATH (only windows)
# read_only_cache = True              # environment CONAN_READ_ONLY_CACHE
# cache_no_locks = True               # environment CONAN_CACHE_NO_LOCKS
# user_home_short = your_path         # environment CONAN_USER_HOME_SHORT
# use_always_short_paths = False      # environment CONAN_USE_ALWAYS_SHORT_PATHS
# skip_vs_projects_upgrade = False    # environment CONAN_SKIP_VS_PROJECTS_UPGRADE
# non_interactive = False             # environment CONAN_NON_INTERACTIVE
# skip_broken_symlinks_check = False  # enviornment CONAN_SKIP_BROKEN_SYMLINKS_CHECK
revisions_enabled = True           # environment CONAN_REVISIONS_ENABLED

# conan_make_program = make           # environment CONAN_MAKE_PROGRAM (overrides the make program used in AutoToolsBuildEnvironment.make)
# conan_cmake_program = cmake         # environment CONAN_CMAKE_PROGRAM (overrides the make program used in CMake.cmake_program)

# cmake_generator                     # environment CONAN_CMAKE_GENERATOR
# cmake_generator_platform            # environment CONAN_CMAKE_GENERATOR_PLATFORM
# http://www.vtk.org/Wiki/CMake_Cross_Compiling
# cmake_toolchain_file                # environment CONAN_CMAKE_TOOLCHAIN_FILE
# cmake_system_name                   # environment CONAN_CMAKE_SYSTEM_NAME
# cmake_system_version                # environment CONAN_CMAKE_SYSTEM_VERSION
# cmake_system_processor              # environment CONAN_CMAKE_SYSTEM_PROCESSOR
# cmake_find_root_path                # environment CONAN_CMAKE_FIND_ROOT_PATH
# cmake_find_root_path_mode_program   # environment CONAN_CMAKE_FIND_ROOT_PATH_MODE_PROGRAM
# cmake_find_root_path_mode_library   # environment CONAN_CMAKE_FIND_ROOT_PATH_MODE_LIBRARY
# cmake_find_root_path_mode_include   # environment CONAN_CMAKE_FIND_ROOT_PATH_MODE_INCLUDE

# msbuild_verbosity = minimal         # environment CONAN_MSBUILD_VERBOSITY

# cpu_count = 1             # environment CONAN_CPU_COUNT

# Change the default location for building test packages to a temporary folder
# which is deleted after the test.
# temp_test_folder = True             # environment CONAN_TEMP_TEST_FOLDER

# cacert_path                         # environment CONAN_CACERT_PATH
# scm_to_conandata                    # environment CONAN_SCM_TO_CONANDATA

# config_install_interval = 1h
# required_conan_version = >=1.26

# keep_python_files = False           # environment CONAN_KEEP_PYTHON_FILES

[storage]
# This is the default path, but you can write your own. It must be an absolute path or a
# path beginning with "~" (if the environment var CONAN_USER_HOME is specified, this directory, even
# with "~/", will be relative to the conan user home, not to the system user home)
path = ./data
# download_cache = /path/to/my/cache

[proxies]
# Empty (or missing) section will try to use system proxies.
# As documented in https://requests.readthedocs.io/en/master/user/advanced/#proxies - but see below
# for proxies to specific hosts
# http = http://user:pass@10.10.1.10:3128/
# http = http://10.10.1.10:3128
# https = http://10.10.1.10:1080
# To specify a proxy for a specific host or hosts, use multiple lines each specifying host = proxy-spec
# http =
#   hostname.to.be.proxied.com = http://user:pass@10.10.1.10:3128
# You can skip the proxy for the matching (fnmatch) urls (comma-separated)
# no_proxy_match = *center.conan.io*, https://myserver.*

[hooks]    # environment CONAN_HOOKS
attribute_checker

# Default settings now declared in the default profile

```

We use this to move the package contents:
```
            for link in ${CONAN_USER_HOME}/.conan/data/*/*/$CONAN_USER/$CONAN_CHANNEL/package/*/.conan_link; do
              source=$(cat $link) && ${MOVE_COMMAND:-mv} $(cygpath "$source")/* $(dirname $link) && rm $link;
            done
```
And we set `CONAN_USER_HOME_SHORT` and `CONAN_USE_ALWAYS_SHORT_PATHS` variables on Windows in addition to `CONAN_USER_HOME`.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @ls-nano 

I am trying to reproduce your use case, but no success, the following seems to capture your description, but works:
```python
    def test_move_package_files(self):
        c = TestClient()
        c.save({"conanfile.py": GenConanfile("pkg", "0.1").with_short_paths(True)})
        c.run("create .")
        c.run("search pkg/0.1@")
        assert "Package_ID: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" in c.out

        ref = ConanFileReference.loads("pkg/0.1")
        pref = PackageReference(ref, NO_SETTINGS_PACKAGE_ID)
        layout = c.cache.package_layout(ref, short_paths=False)
        pkg_folder = layout.package(pref)
        link = os.path.join(pkg_folder, ".conan_link")
        link_folder = load(link)
        for f in os.listdir(link_folder):
            shutil.copy2(os.path.join(link_folder, f), pkg_folder)
        os.remove(link)

        c.run("search pkg/0.1@")
        assert "Package_ID: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" in c.out
```

In any case, the problem is that modifying the Conan cache is undefined behavior (this has been made explicit in https://docs.conan.io/2/knowledge/guidelines.html), and specially once revisions are defined, Conan packages have to be immutable. It is very likely that that movement of file is violating some of the assumptions, and looks like that it won't be possible to change that in 1.X scope.


>  This is so that we can upload and move the package around without having to fear breaking the links.

Probably this is what needs more clarification. It is not necessary to move the files to be able to upload the package, install the package, etc., everything should work normally without needing this extra step.

**Comment by ls-nano** [comment]
Thank you for you reply and excuse me for the delay

> This is so that we can upload and move the package around without having to fear breaking the links.

We usually tarball the package after building it on CI for exporting it as a artifact for manual testing in some cases, which is why we used the code above to move the package contents. I've now moved the tarballing step after the upload, so that we don't break the package for uploading (it might still be broken in the tarball, but thats not too big of an issue). Thanks again!
